### PR TITLE
Add measurement example involving NI-DCPower and NI-VISA DMM instruments

### PR DESCRIPTION
**Pull request by samchris007**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/measurement-services-python/blob/main/CONTRIBUTING.md).

### What does this Pull Request accomplish?
This measurement example explains handling two different instruments namely an NI instrument(NI-DCPower) and a third-party instrument(NI-VISA DMM), within a single measurement service.
-  This example also defines creating sessions(reserving and registering sessions) for two completely different instruments within a single measurement service.

### Implementation
Add measurement example which involves sourcing NI-DCPower to supply an input pin and measuring the output pin using the NI-VISA DMM instrument and an NI Instrument Simulator v2.0..

- Implemented measure function by replicating sourcing and measuring logic from existing examples [nidcpower_source_dc_voltage](https://github.com/ni/measurementlink-python/tree/main/examples/nidcpower_source_dc_voltage) and [nivisa_dmm_measurement](https://github.com/ni/measurementlink-python/tree/main/examples/nivisa_dmm_measurement).
- For reserving and creating sessions,
  - Both input and output pins are reserved using `session_management_client.reserve_sessions()`
  - Session objects for both NI-DCPower and NI-VISA DMM instruments are created within a single `with` statement, 

### What testing has been done?

Manually tested with both simulation and hardware.

### CODE REVIEW

File: examples/nidcpower_nivisa_dmm_measurement/measurement.py
Comment (keer1111) [inline examples/nidcpower_nivisa_dmm_measurement/measurement.py:180]: will this list be empty? should be checked?

File: examples/nidcpower_nivisa_dmm_measurement/_helpers.py
Comment (keer1111) [inline examples/nidcpower_nivisa_dmm_measurement/_helpers.py:271]: public methods can have doc string?

**Comment by bkeryan** [comment]
I really think we should revisit https://github.com/ni/measurementlink-python/pull/256

File: examples/nidcpower_nivisa_dmm_measurement/NIDCPowerNIVisaDmmMeasurement.instudioproj
Comment (bkeryan) [inline examples/nidcpower_nivisa_dmm_measurement/NIDCPowerNIVisaDmmMeasurement.instudioproj:1]: This is missing TestStand support.

When you add it, please don't combine the NI-DCPower and VISA code into a single, monolithic `teststand_fixture.py` module. Use separate modules for pin map, NI-DCPower, and VISA so that we can "vendor" identical copies of these modules in multiple examples. I previously started to do this in #256, but I put it on hold due to the other example refactoring.

File: examples/nidcpower_nivisa_dmm_measurement/pyproject.toml
Comment (bkeryan) [inline examples/nidcpower_nivisa_dmm_measurement/pyproject.toml:2]: I don't like this example name. It sounds like it's using both NI-DCPower and NI-VISA to do a DMM measurement, but NI-DCPower is not involved in the DMM measurement.

What is it really doing? 

It's using NI-DCPower to source a DC voltage for the DUT's input pin, and it's using VISA and a DMM to measure the DUT's output pin.

Is there a higher level way to describe this? 

"Stimulus response" comes to mind, but I think that might imply time domain signals, whereas this is working with DC signals.

What sort of DUT would you test with this measurement?

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: examples/nidcpower_nivisa_dmm_measurement/measurement.py
Comment (samchris007) [reply examples/nidcpower_nivisa_dmm_measurement/measurement.py:58]: I was looking into one of the LabVIEW examples for Agilent 34401 and I noticed that it has the same property name of `Function` for configuring the measurement. So, I think it's ok to use `Function` here in reference to the LabVIEW example.

image

Please share your thoughts on this.

File: examples/nidcpower_nivisa_dmm_measurement/_helpers.py
Comment (samchris007) [reply examples/nidcpower_nivisa_dmm_measurement/_helpers.py:1]: Ok. Updated the helpers to the original examples.
Changes made in helpers
- Renamed APIs `create_visa_resource_manager`, `create_visa_session` to `create_resource_manager`, `create_session` in **_visa_helpers.py**.
- Move `USE_SIMULATION` to measurement.py so as to normalize the simulation between two instruments in both the instrument helpers.

File: examples/nidcpower_nivisa_dmm_measurement/_helpers.py
Comment (samchris007) [reply examples/nidcpower_nivisa_dmm_measurement/_helpers.py]: Sure. Updated the session creation APIs to use `service_options.use_simulation` as suggested.

File: examples/nidcpower_nivisa_dmm_measurement/measurement.py
Comment (samchris007) [reply examples/nidcpower_nivisa_dmm_measurement/measurement.py]: Oh sure. 
- Renamed APIs `create_visa_resource_manager` and `create_visa_session` to `create_resource_manager` and `create_session` in **_visa_helpers.py**.
- Updated the example using module scoping.

File: examples/nidcpower_nivisa_dmm_measurement/pyproject.toml
Comment (samchris007) [reply examples/nidcpower_nivisa_dmm_measurement/pyproject.toml]: As discussed in our PO Connect, I've renamed the measurement to `Output Voltage Measurement`.

Also, I will be creating a separate PR for moving this example into the **multi_instrument** sub-directory with the necessary workflow changes.

File: examples/nidcpower_nivisa_dmm_measurement/measurement.py
Comment (samchris007) [reply examples/nidcpower_nivisa_dmm_measurement/measurement.py:1]: As discussed in our PO Connect, I have updated the order of measurement operations,
1. Instead of using the trigger, configured both instruments at the start.
2. Initiated the SMU and waited for the source complete.
3. Then, Read the DMM

File: examples/nidcpower_nivisa_dmm_measurement/measurement.py
Comment (samchris007) [reply examples/nidcpower_nivisa_dmm_measurement/measurement.py]: We are aligned that there is no or less advantage in triggering the DMM as both are different instruments and are not connected through the same PXI. 
> If you are trying to minimize test time, then you should configure both instruments up front, then initiate the SMU, wait for source complete, and read the DMM. This way, the SMU's [source delay property](https://www.ni.com/docs/en-US/bundle/ni-dcpower-properties/page/dcpowerpropref/pnidcpower_sourcedelay.html) more accurately controls the time between the source and measurement operations.

As suggested, I have updated the order of the measurement operations by configuring both instruments at the start, initiating the SMU and waiting for the source complete, and then reading the DMM.

So please share your thoughts on whether we still need to return the SMU's InPin measurements.

File: examples/nidcpower_nivisa_dmm_measurement/measurement.py
Comment (samchris007) [reply examples/nidcpower_nivisa_dmm_measurement/measurement.py]: `session_info` will only be empty if the existing code gets edited.
Hence, Not handling empty lists should be okay in this context.

File: examples/nidcpower_source_dc_voltage/_nidcpower_helpers.py
Comment (bkeryan) [inline examples/nidcpower_source_dc_voltage/_nidcpower_helpers.py:13]: `teststand_fixture.py` is still using the old signature.

https://github.com/ni/measurementlink-python/blob/main/examples/nidcpower_source_dc_voltage/teststand_fixture.py#L63
```
                _ = create_session(
                    session_info,
                    grpc_device_channel,
                    initialization_behavior=nidcpower.SessionInitializationBehavior.INITIALIZE_SERVER_SESSION,
                )
```

File: examples/nivisa_dmm_measurement/teststand_fixture.py
Comment (bkeryan) [inline examples/nivisa_dmm_measurement/teststand_fixture.py:39]: If we are going to move `use_simulation` to the TestStand file globals:
- Do it for all examples in a separate PR
- Update README.md

Comment (bkeryan) [inline examples/nivisa_dmm_measurement/teststand_fixture.py:39]: @rfriedma I think you originally suggested the `USE_SIMULATION` constant. Is moving it to the TestStand file globals okay with you?

File: examples/output_voltage_measurement/NIInstrumentSimulatorV2_0.yaml
Comment (bkeryan) [inline examples/output_voltage_measurement/NIInstrumentSimulatorV2_0.yaml:1]: The path we discussed in our conference call on Thursday was `examples/multi_instrument/output_voltage_measurement/`

File: examples/output_voltage_measurement/README.md
Comment (bkeryan) [inline examples/output_voltage_measurement/README.md:3]: as input to what?
measures output of what?

Use the word "DUT".

File: examples/output_voltage_measurement/README.md
Comment (bkeryan) [inline examples/output_voltage_measurement/README.md:7]: don't combine these

File: examples/output_voltage_measurement/README.md
Comment (bkeryan) [inline examples/output_voltage_measurement/README.md:11]:
```suggestion
- Pin-aware, supporting one session and pin per instrument and a single site
  - Sources the DC voltage level of the DUT input pin
  - Measures the voltage of the DUT output pin
```

File: examples/output_voltage_measurement/README.md
Comment (bkeryan) [inline examples/output_voltage_measurement/README.md:3]: "NI-VISA DMM" may be useful shorthand but it's imprecise at best and misleading at worst. That's why the other readme doesn't use this term.

File: examples/output_voltage_measurement/README.md
Comment (bkeryan) [inline examples/output_voltage_measurement/README.md:31]: but in TestStand...

File: examples/nivisa_dmm_measurement/measurement.py
Comment (bkeryan) [inline examples/nivisa_dmm_measurement/measurement.py:101]: FYI, we could remove another call by making `create_session`'s `resource_manager` parameter optional. However, I guess we would have to pass `use_simulation` instead.

File: examples/output_voltage_measurement/teststand_nivisadmm.py
Comment (bkeryan) [inline examples/output_voltage_measurement/teststand_nivisadmm.py:1]: `teststand_nivisa_dmm.py`

File: examples/output_voltage_measurement/README.md
Comment (bkeryan) [inline examples/output_voltage_measurement/README.md:3]: I thought we were going to add content about how this example also works with a real DMM, like a 34401A

File: examples/output_voltage_measurement/pyproject.toml
Comment (bkeryan) [inline examples/output_voltage_measurement/pyproject.toml:9]:
```suggestion
ni-measurementlink-service = {version = "^1.1.0"}
```

File: examples/output_voltage_measurement/teststand_pin_map.py
Comment (bkeryan) [inline examples/output_voltage_measurement/teststand_pin_map.py:1]: Are you going to do this for all the examples? in a separate PR?

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: examples/output_voltage_measurement/NIInstrumentSimulatorV2_0.yaml
Comment (samchris007) [reply examples/output_voltage_measurement/NIInstrumentSimulatorV2_0.yaml:1]: Yes. That's right. But we must also need to update the GitHub workflows(**check_examples.yml** and **install_examples.py**). Hence, the moving of `output_voltage_measurement` into the `multi_instrument` sub-directory will be raised as a separate PR.

File: examples/nivisa_dmm_measurement/teststand_fixture.py
Comment (rfriedma) [reply examples/nivisa_dmm_measurement/teststand_fixture.py:39]: I lean yes however, if a user sees "use_simulation" as a global on the sequence... I think it'd be expected that it would simulate everything. Would that be the case?

Comment (rfriedma) [reply examples/nivisa_dmm_measurement/teststand_fixture.py:39]: I agree with that approach for all examples that support simulation. Add a new File Global as sibling to the MeasurementLink container (not within it).

File: examples/nidcpower_source_dc_voltage/_nidcpower_helpers.py
Comment (Jayaseelan-James) [reply examples/nidcpower_source_dc_voltage/_nidcpower_helpers.py]: I have reverted the changes specific to `nidcpower` and `nivisa_dmm` measurement. Will take this up as part of work item 2473562.

File: examples/nivisa_dmm_measurement/teststand_fixture.py
Comment (Jayaseelan-James) [reply examples/nivisa_dmm_measurement/teststand_fixture.py]: Resolving the conversation now, but the change will be taken up as part of work item 2473562.

File: examples/output_voltage_measurement/README.md
Comment (Jayaseelan-James) [reply examples/output_voltage_measurement/README.md]: Updated to `... measures the DUT output with DMM using NI-VISA.`.

File: examples/output_voltage_measurement/README.md
Comment (Jayaseelan-James) [reply examples/output_voltage_measurement/README.md]: Added individual points explaining the edit flow for both MeasurementLink UI Editor / InstrumentStudio and TestStand.

File: examples/nivisa_dmm_measurement/measurement.py
Comment (Jayaseelan-James) [reply examples/nivisa_dmm_measurement/measurement.py]: Updated this change for `output_voltage_measurement` , whereas updates to the `nivisa_dmm_measurement` will be taken up as part of work item 2473562.

File: examples/output_voltage_measurement/README.md
Comment (Jayaseelan-James) [reply examples/output_voltage_measurement/README.md]: Added the below points to emphasize the use of real DMM:

To use a DMM like Keysight 34401A:
  - Update the [`OutputVoltageMeasurement.pinmap`](./OutputVoltageMeasurement.pinmap) to reflect the connected instrument's type and resource name.
  - Additionally, verify driver specific commands in `_visa_helpers.py` and update them when required.

File: examples/nivisa_dmm_measurement/teststand_fixture.py
Comment (Jayaseelan-James) [reply examples/nivisa_dmm_measurement/teststand_fixture.py]: @rfriedma, @bkeryan Now, I have added a new `_constants.py` file under `output_voltage_measurement` directory which holds the `USE_SIMULATION` boolean and is imported in `measurement.py`, `teststand_nidcpower.py` and `teststand_nivisa_dmm.py` and passed to the `create_session` method. 

Pros of this approach:
1. The existing user flow of updating the USE_SIMULATION bool in a .py is retained, just that now the user has to update the instrument mode in `_constants.py` instead of `_<instrument>_helpers.py`.
2. A `UseSimulation` boolean in teststand sequence is not required as it is handled using the same boolean in `_constants.py`. So, the user will update only one file to switch between hardware and simulation instead of updating in .py and in TestStand.

Cons:
1. A new `_constants.py` file is added.

File: examples/output_voltage_measurement/README.md
Comment (bkeryan) [inline examples/output_voltage_measurement/README.md:8]: This does not match https://github.com/ni/measurementlink-python/blob/main/examples/nivisa_dmm_measurement/README.md
```suggestion
- Uses the open-source `PyVISA` package to access NI-VISA from Python
```

File: examples/output_voltage_measurement/README.md
Comment (bkeryan) [inline examples/output_voltage_measurement/README.md:41]: This is out of date

File: examples/output_voltage_measurement/README.md
Comment (bkeryan) [inline examples/output_voltage_measurement/README.md:56]:
```suggestion
To use a 3rd party DMM, such as a Keysight 34401A:
```

File: examples/output_voltage_measurement/README.md
Comment (bkeryan) [inline examples/output_voltage_measurement/README.md:58]:
```suggestion
  - Additionally, verify driver specific commands in `_visa_helpers.py` and update them if required.
```

File: examples/output_voltage_measurement/README.md
Comment (bkeryan) [inline examples/output_voltage_measurement/README.md:56]: Add this paragraph to the other VISA example's readme.

File: examples/output_voltage_measurement/README.md
Comment (bkeryan) [inline examples/output_voltage_measurement/README.md:57]: Changing the instrument type in the pinmap requires changing the Python code and the .sfp file, so let's not recommend that here.

File: examples/output_voltage_measurement/_constants.py
Comment (bkeryan) [inline examples/output_voltage_measurement/_constants.py:2]: It's weird that ni-python-styleguide didn't make you leave a blank line here.

File: examples/output_voltage_measurement/measurement.py
Comment (bkeryan) [inline examples/output_voltage_measurement/measurement.py:125]: This should be obvious from the code.

File: examples/output_voltage_measurement/measurement.py
Comment (bkeryan) [inline examples/output_voltage_measurement/measurement.py:184]: This doesn't work due to Python scoping rules. Unfortunately, this minor refactor seems to have opened another "can of worms" that may lead to changing all of the examples.

The nested function `cancel_callback` can modify the `pending_cancellation` variable captured from the `_add_cancel_callback` function. However, this has no effect on the `measure` function because `bool` in Python behaves as if it is passed by value.

Really, it's more complicated: there are two singleton instances of `bool`: `True` and `False`, and you pass around references to them. However, these instances are immutable, so assigning to a `bool` variable works by rebinding it to a different reference. `_add_cancel_callback` cannot modify `measure`'s `pending_cancellation` variable.

For more details and explanation, read:
- https://realpython.com/python-scope-legb-rule/
- https://stackoverflow.com/questions/1494696/are-pythons-bools-passed-by-value

An alternative is to use `threading.Event` like in https://github.com/ni/measurementlink-python/blob/main/examples/ui_progress_updates/measurement.py . Event objects are mutable, so you can have a callback set the event without rebinding any references. As a plus, you don't even need to define a callback function, because you can use the event's `set` method as the callback.

In addition, calling `abort()` on the NI-DCPower session doesn't make as much sense as we originally thought. It doesn't unblock waiting API calls like it does in NI-DAQmx. Also, raising an exception in the `with session.initiate()` block automatically calls `abort()`, so we don't really need to do it again.

File: examples/output_voltage_measurement/pyproject.toml
Comment (bkeryan) [inline examples/output_voltage_measurement/pyproject.toml:20]: sorry, this is wrong in all the examples

```suggestion
# nidcpower = { git = "https://github.com/ni/nimi-python.git", subdirectory = "generated/nidcpower", extras = ["grpc"]}
```

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: examples/output_voltage_measurement/_constants.py
Comment (Jayaseelan-James) [reply examples/output_voltage_measurement/_constants.py]: Included a blank line now.

File: examples/output_voltage_measurement/pyproject.toml
Comment (Jayaseelan-James) [reply examples/output_voltage_measurement/pyproject.toml]: Sure, I updated this example's `pyproject.toml` now. Will update the other examples as a part of work item 2484653.

File: examples/output_voltage_measurement/README.md
Comment (Jayaseelan-James) [reply examples/output_voltage_measurement/README.md]: Will update this as part of work item 2473562.

File: examples/output_voltage_measurement/measurement.py
Comment (Jayaseelan-James) [reply examples/output_voltage_measurement/measurement.py]: I have updated the `measurement.py` to use `threading.Event` for "cancellation_callback".

File: examples/output_voltage_measurement/README.md
Comment (Jayaseelan-James) [reply examples/output_voltage_measurement/README.md]: Updated.

File: examples/output_voltage_measurement/measurement.py
Comment (bkeryan) [inline examples/output_voltage_measurement/measurement.py:186]: The purpose of this function is to wait for the source complete event. Ideally, we would call `channels.wait_for_event()` once with a long timeout, but there is no way to cancel this long wait, so we break it up into shorter 100 ms waits and check for cancellation after each short wait.

Adding a second wait to the loop hurts performance because it will keep waiting for up to 100 ms after the source complete event has occurred.

`channels.wait_for_event()` performs a synchronous wait and doesn't support multiplexed I/O like `select()` or `WaitForMultipleObjects()`, so the best solution is to call `channels._wait_for_event()` with a 100 ms timeout and use the cancellation event as an atomic flag.

```suggestion
        if cancellation_event.is_set():
```

**Review by bkeryan — APPROVED** [review]
Approved with suggestions:
- Use `event.is_set()` instead of `event.wait(timeout=0.1)`

File: examples/output_voltage_measurement/measurement.py
Comment (Jayaseelan-James) [reply examples/output_voltage_measurement/measurement.py]: Thanks for the suggestion. Updated to invoke `is_set()`.

**Comment by Jayaseelan-James** [comment]
@dixonjoel, @csjall Please help review this PR.

**Review by dixonjoel — APPROVED** [review]

**Comment by dixonjoel** [comment]
> @dixonjoel, @csjall Please help review this PR.

Done. I don't think you need to wait for @csjall

**Pull request merged**
